core/src/main/java/org/apache/iceberg/view/BaseViewOperations.java

```diff
@@ -0,0 +1,215 @@
+  }
+
+  protected void doRefresh() {
+    throw new UnsupportedOperationException("Not implemented: doRefresh");
+  }
+
+  protected void doCommit(ViewMetadata base, ViewMetadata metadata) {
+    throw new UnsupportedOperationException("Not implemented: doCommit");
+  }
+
+  protected abstract String viewName();
+
+  protected abstract FileIO io();
+
+  public int currentVersion() {
+    return version;
+  }
+
+  public String currentMetadataLocation() {
+    return currentMetadataLocation;
+  }
+
+  @Override
+  public ViewMetadata current() {
+    if (shouldRefresh) {
+      return refresh();
+    }
+
+    return currentMetadata;
+  }
+
+  @Override
+  public ViewMetadata refresh() {
+    boolean currentMetadataWasAvailable = currentMetadata != null;
+    try {
+      doRefresh();
+    } catch (NoSuchViewException e) {
+      if (currentMetadataWasAvailable) {
+        LOG.warn("Could not find the view during refresh, setting current metadata to null", e);
+        shouldRefresh = true;
+      }
+
+      currentMetadata = null;
+      currentMetadataLocation = null;
+      version = -1;
+      throw e;
+    }
+
+    return current();
+  }
+
+  @Override
+  public void commit(ViewMetadata base, ViewMetadata metadata) {
+    // if the metadata is already out of date, reject it
+    if (base != current()) {
+      if (base != null) {
+        throw new CommitFailedException("Cannot commit: stale view metadata");
+      } else {
+        // when current is non-null, the view exists. but when base is null, the commit is trying
+        // to create the view
+        throw new AlreadyExistsException("View already exists: %s", viewName());
+      }
+    }
+
+    // if the metadata is not changed, return early
+    if (base == metadata) {
+      LOG.info("Nothing to commit.");
+      return;
+    }
+
+    long start = System.currentTimeMillis();
+    doCommit(base, metadata);
+    requestRefresh();
+
+    LOG.info(
+        "Successfully committed to view {} in {} ms",
+        viewName(),
+        System.currentTimeMillis() - start);
+  }
+
+  protected String writeNewMetadata(ViewMetadata metadata, int newVersion) {
+    String newMetadataFilePath = newMetadataFilePath(metadata, newVersion);
+    OutputFile newMetadataLocation = io().newOutputFile(newMetadataFilePath);
+
+    // write the new metadata
+    // use overwrite to avoid negative caching in S3. this is safe because the metadata location is
+    // always unique because it includes a UUID.
+    ViewMetadataParser.overwrite(metadata, newMetadataLocation);
+
+    return newMetadataLocation.location();
+  }
+
+  private String newMetadataFilePath(ViewMetadata metadata, int newVersion) {
+    return metadataFileLocation(
+        metadata, String.format("%05d-%s%s", newVersion, UUID.randomUUID(), ".metadata.json"));
+  }
+
+  private String metadataFileLocation(ViewMetadata metadata, String filename) {
+    return String.format("%s/%s/%s", metadata.location(), METADATA_FOLDER_NAME, filename);
```

Comment:
Where do we ensure that `metadata.location()` does not have a trailing slash? I think we need to call `LocationUtil.stripTrailingSlash`.

Reply:
the equivalent code in `BaseMetastoreTableOperations` seemed to be using `LocationUtil.stripTrailingSlash` only for the case when `TableProperties.WRITE_METADATA_LOCATION` was set, which we currently don't have here.
Related to this, should we maybe have a similar property in `ViewProperties`?

Reply:
I've added `ViewProperties.WRITE_METADATA_LOCATION` and also called `LocationUtil.stripTrailingSlash`